src/core/mem.c

```diff
@@ -158,39 +181,83 @@ static bool root_pool_set_up_bitmap(paddr_t load_addr, struct page_pool* root_po
     return mem_reserve_ppool_ppages(root_pool, &bitmap_pp);
 }
 
-static bool pp_root_reserve_hyp_mem(paddr_t load_addr, struct page_pool* root_pool)
+static bool pp_reserve_hyp_image_load(struct page_pool* pool)
+{
+    size_t image_load_size = (size_t)(&_image_load_end - &_image_start);
+
+    struct ppages images_load_ppages = mem_ppages_get(img_addr, NUM_PAGES(image_load_size));
+
+    return mem_reserve_ppool_ppages(pool, &images_load_ppages);
+}
+
+static bool pp_reserve_hyp_image_noload(struct page_pool* pool)
 {
     size_t image_load_size = (size_t)(&_image_load_end - &_image_start);
     size_t image_noload_size = (size_t)(&_image_end - &_image_load_end);
     size_t vm_image_size = (size_t)(&_vm_image_end - &_vm_image_start);
-    size_t cpu_size = platform.cpu_num * mem_cpu_boot_alloc_size();
-    paddr_t image_noload_addr = load_addr + image_load_size + vm_image_size;
-    paddr_t cpu_base_addr = image_noload_addr + image_noload_size;
+    paddr_t image_noload_addr = img_addr + image_load_size + vm_image_size;
 
-    struct ppages images_load_ppages = mem_ppages_get(load_addr, NUM_PAGES(image_load_size));
     struct ppages images_noload_ppages =
         mem_ppages_get(image_noload_addr, NUM_PAGES(image_noload_size));
+
+    return mem_reserve_ppool_ppages(pool, &images_noload_ppages);
+}
+
+static bool pp_reserve_cpus(struct page_pool* pool)
+{
+    size_t image_load_size = (size_t)(&_image_load_end - &_image_start);
+    size_t image_noload_size = (size_t)(&_image_end - &_image_load_end);
+    size_t vm_image_size = (size_t)(&_vm_image_end - &_vm_image_start);
+    size_t cpu_size = platform.cpu_num * mem_cpu_boot_alloc_size();
+    paddr_t image_noload_addr = img_addr + image_load_size + vm_image_size;
+    paddr_t cpu_base_addr = image_noload_addr + image_noload_size;
+
+    struct ppages cpu_ppages = mem_ppages_get(cpu_base_addr, NUM_PAGES(cpu_size));
+
+    return mem_reserve_ppool_ppages(pool, &cpu_ppages);
+}
+
+static bool pp_reserve_hyp_data(struct page_pool* root_pool)
+{
+    size_t data_size = (size_t)(&_image_end - &_data_vma_start);
+    size_t cpu_size = platform.cpu_num * mem_cpu_boot_alloc_size();
+    paddr_t data_base_addr = (paddr_t)&_data_vma_start;
+    paddr_t cpu_base_addr = data_base_addr + data_size;
```

Comment:
How is this guaranteed to be aligned to PAGE_SIZE?

Reply:
Done 04f6748a